File: DailyChallenges/Scala/LeetCode_1011.scala
object Solution {
    def shipWithinDays(weights: Array[Int], days: Int): Int = {
        var left = weights.reduceLeft(_ max _)
        var right = 25000000
        while (left < right) {
            val mid = (left+right)/2
            val need = weights.foldLeft((0, 1)){(p, w) => if (w + p._1 <= mid) (p._1+w, p._2) else (w, p._2+1)}._2
            if (need > days) left = mid+1 else right = mid
        }
        left
    }
}

File: DailyChallenges/CPP/LeetCode_1011.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int days) {
       function<bool(int)> check = [&](int w){
           int d = 0, i = 0;
           while (i < weights.size()){
               int wd = 0;
               while (wd < w && i <weights.size()){
                   if (wd + weights[i] > w) break;
                   else wd += weights[i++];
               }
               if (++d>days) return false;
           }
           return true;
        };

       int l = 0; int r = 0;
       for(int w:weights){
           l = max(l, w);r += w;
       }

       while (l<r){
            int m = (l + r) / 2;
            if (check(m)) r=m; else l=m+1;
       }

       return l;
    }
};
